whitepaper/wlroots.tex

```diff
@@ -0,0 +1,440 @@
+paradigm. It is also less modular and difficult to break into several pieces.
+Additionally, as a consequence of being refactored out of an existing
+compositor, it inherits many design decisions of Weston itself, which may not be
+amenable to other compositors.
+
+\textbf{swc} and \textbf{wlc} are similar. They both strike a reasonable balance
+between size and features, but both make decisions for the compositor that lock
+them into particular designs and make novel compositors more difficult to
+write. The advantage of their approach is that it is easy to use them to build a
+simple working compositor - in a matter of minutes, not days or weeks.
+
+Another approach is to simply write your compositor entirely in-house. This is
+the approach favored by \textbf{mutter} and \textbf{kwin} of the GNOME and KDE
+projects respectively. The advantage of this approach is the flexibility to
+design your compositor in any way you please. However, lots of work is
+duplicated between projects with this approach, which can lead to compositors
+which are subtly incompatible with each other.
+
+The approach taken by \textbf{wlroots} offers reusability and consistency
+between compositors based on it, and offloads lots work from the compositor
+implementation. This is accomplished without making design decisions on behalf
+of the compositor implementation: we believe virtually any novel kind of
+compositor can be built on top of wlroots.
+
+The main disadvantage of our approach is that batteries are not included. More
+work is required to get a working compositor compared to libweston or wlc.
+However, you will not hit the limitations of these libraries, and will more
+quickly arrive at a diverse featureset than with the kwin or mutter approaches.
+
+\section{High-level Architecture}\label{highlevel}
+
+wlroots is organized into many discrete components that can be used separately
+or replaced with your own implementations, and freely composed.
+
+\subsection{Backends}\label{high level backends}
+
+Backends are used to abstract access to input and output hardware. A backend,
+once initialized, provides some number of \textbf{wlr_outputs} and
+\textbf{wlr_input_devices}, the latter being an abstraction on top of types
+like \textbf{wlr_pointer}, \textbf{wlr_keyboard}, \textbf{wlr_touch}, and so on.
+
+Backends include \emph{drm} and \emph{libinput}, which abstract the
+corresponding Linux subsystems; \emph{wayland}, which uses Wayland windows on
+another running Wayland compositor; the similar \emph{x11} backend; \emph{rdp},
+which uses the remote desktop protocol; and so on. An interface is provided so
+that compositor implementations can supply their own backend implementations as
+well.
+
+Also available is the \emph{multi} backend, which acts as a container for many
+backends. This is necessary, for example, to use DRM and libinput together. It
+also allows you to add and remove backends at runtime. An example use-case:
+adding an RDP backend at runtime.
```

Comment:
Consistency here: backends given earlier were not capitalized (and are not in the code, event if the emphasis on lower case x11 looks odd when rendered), but are capitalized here without emphasis.